src/github.com/mozilla-services/pushgo/simplepush/handlers_endpoint.go

```diff
@@ -304,35 +309,53 @@ sendUpdate:
 		}
 		status, _ := ErrToStatus(err)
 		h.metrics.Increment("updates.appserver.error")
-		http.Error(resp, "Could not update channel version", status)
+		writeJSON(resp, status, []byte(`"Could not update channel version"`))
 		return
 	}
 
-	// Ping the appropriate server
-	// Is this ours or should we punt to a different server?
+	cn, _ := resp.(http.CloseNotifier)
+	if !h.deliver(cn, uaid, chid, version, requestID, data) {
+		writeJSON(resp, http.StatusNotFound, []byte("false"))
+		return
+	}
+
+	writeSuccess(resp)
+	updateSent = true
+	return
+}
+
+// deliver routes an incoming update to the appropriate server.
+func (h *EndpointHandler) deliver(cn http.CloseNotifier, uaid, chid string, version int64, requestID string, data string) (ok bool) {
 	client, clientConnected := h.app.GetClient(uaid)
-	if !clientConnected {
-		// TODO: Move PropPinger here? otherwise it's connected?
+	// Always route to other servers first, in case we're holding open a stale
+	// connection and the client has already reconnected to a different server.
+	if h.alwaysRoute || !clientConnected {
 		h.metrics.Increment("updates.routed.outgoing")
+		// Abort routing if the connection goes away.
 		var cancelSignal <-chan bool
-		if cn, ok := resp.(http.CloseNotifier); ok {
+		if cn != nil {
 			cancelSignal = cn.CloseNotify()
 		}
-		if err = h.router.Route(cancelSignal, uaid, chid, version, time.Now().UTC(), requestID, data); err != nil {
-			resp.WriteHeader(http.StatusNotFound)
-			resp.Write([]byte("false"))
-			return
+		// Route the update.
+		ok, _ = h.router.Route(cancelSignal, uaid, chid, version,
+			timeNow().UTC(), requestID, data)
+		if ok {
+			return true
 		}
 	}
-
-	if clientConnected {
-		h.app.Server().RequestFlush(client, chid, int64(version), data)
-		h.metrics.Increment("updates.appserver.received")
+	// If the device is not connected to this server, indicate whether routing
+	// was successful.
+	if !clientConnected {
+		return ok
```

Comment:
Sounds good. This function didn't use bare returns for the other three cases, so I used an explicit one here for consistency.